bot/cogs/whitelist.py

```diff
@@ -0,0 +1,113 @@
+import logging
+
+from discord import Colour, Embed
+from discord.ext.commands import Bot, Cog, Context, group
+
+from bot.constants import Roles
+from bot.decorators import with_role
+from bot.pagination import LinePaginator
+
+log = logging.getLogger(__name__)
+
+
+class Whitelist(Cog):
+    """Save new whitelists and fetch existing whitelist."""
+
+    def __init__(self, bot: Bot):
+        self.bot = bot
+
+    async def _get_whitelisted_items(self, whitelist_type: str) -> list:
+        """Returns the list of current whitelisted items for the type."""
+        query_params = {"type": whitelist_type}
+        result = await self.bot.api_client.get(f"bot/whitelist", params=query_params)
+        whitelisted_items = [item["whitelisted_item"] for item in result]
+        return whitelisted_items
+
+    @group(name="whitelist", invoke_without_command=True)
+    @with_role(Roles.owners, Roles.admins)
+    async def whitelist_group(
+        self, ctx: Context, whitelist_type: str, *, whitelist_item: str
+    ) -> None:
+        """Commands for managing your whitelist."""
+        await ctx.invoke(
+            self.set_command,
+            whitelist_type=whitelist_type,
+            whitelist_item=whitelist_item,
+        )
+
+    @whitelist_group.command(name="add", aliases=("append", "a"))
+    @with_role(Roles.owners, Roles.admins)
+    async def set_command(
+        self, ctx: Context, whitelist_type: str, *, whitelist_item: str
+    ) -> None:
+        """Add a new item to the whitelist for the whitelist_type."""
+        whitelist_type, whitelist_item = (
+            whitelist_type.lower().strip(),
+            whitelist_item.strip(),
+        )
+        existing_whitelist = await self._get_whitelisted_items(whitelist_type)
+        if whitelist_item in existing_whitelist:
+            await ctx.send(
+                embed=Embed(
+                    description=f"**{whitelist_item} already exists in the {whitelist_type} whitelist.**",
+                    colour=Colour.red(),
+                )
+            )
+        else:
+            body = {"type": whitelist_type, "whitelisted_item": whitelist_item}
+            await self.bot.api_client.post("bot/whitelist", json=body)
+
+            log.debug(
+                f"{ctx.author} successfully added the following items to our database: \n"
+                f"whitelist_type: {whitelist_type}\n"
+                f"whitelist_item: '{whitelist_item}'\n"
+            )
+
+            await ctx.send(
+                embed=Embed(
+                    title="Item successfully added",
+                    description=f"**{whitelist_item}** added to {whitelist_type} whitelist database.",
+                    colour=Colour.blurple(),
+                )
+            )
+
+    @whitelist_group.command(name="get", aliases=("show", "g"))
+    @with_role(Roles.owners, Roles.admins)
```

Comment:
I think it'd make sense to at least let mods (and maybe helpers?) see what the whitelist contains. After all, this is possible with the current system since the config is public in the repo. Will need input from other staff on this.